### PR TITLE
[SPARK-22881][ML][TEST] ML regression package testsuite add StructuredStreaming test

### PR DESCRIPTION
**Pull request by WeichenXu123**
## What changes were proposed in this pull request?

ML regression package testsuite add StructuredStreaming test

In order to make testsuite easier to modify, new helper function added in `MLTest`:
```
def testTransformerByGlobalCheckFunc[A : Encoder](
      dataframe: DataFrame,
      transformer: Transformer,
      firstResultCol: String,
      otherResultCols: String*)
      (globalCheckFunction: Seq[Row] => Unit): Unit
```

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by WeichenXu123** [comment]
@MrBago @jkbradley

**Comment by jkbradley** [comment]
Thanks!  Let's track these tasks in new JIRAs.  I made one for regression just now: https://issues.apache.org/jira/browse/SPARK-22881

**Comment by MrBago** [comment]
@WeichenXu123 it looks like `testTransformer` is a special case of `testTransformerByGlobalCheckFunc`. I think it's cleaner to structure the tests in this way instead of passing around nulls, because of the time crunch I made a PR to show what I mean, https://github.com/WeichenXu123/spark/pull/3.

**Comment by jkbradley** [comment]
Actually, going further than what Bago said: All of the places which use globalCheckFunction assume that Dataset.collect() returns the Rows in a fixed order.  We should really fix those unit tests to check values row-by-row.  As a side effect, that would allow us to eliminate globalCheckFunction.

**Comment by WeichenXu123** [comment]
@jkbradley
There're two cases which can use `globalCheckFunction`
- test statistics (such as min/max ) on global transformer output
- get global result array and compare it with hardcoding array values.

If remove `globalCheckFunction`, tests in above cases will be harder to write. So I prefer to reserve `globalCheckFunction`. What do you think ?

**Comment by WeichenXu123** [comment]
@MrBago Merge your code suggestion. Thanks!

**Comment by jkbradley** [comment]
> assume that Dataset.collect() returns the Rows in a fixed order.

I'm quite sure that:
* When the Dataset has been constructed without any shuffles or repartitions, then Rows are always in a fixed order.
* When there has been a shuffle, it is likely the Rows will not follow a fixed order.
* Spark APIs never guarantee a fixed order, except when sorting has been performed.

Basically, we should try to avoid design patterns which assume fixed Row orders.  It may be safe sometimes, but the assumption can lead to mistakes.

> There're two cases which can use globalCheckFunction
>  * test statistics (such as min/max ) on global transformer output
>  * get global result array and compare it with hardcoding array values.

For test statistics, globalCheckFunction makes sense.
* But none of the tests in this PR require this.  Are there any unit tests in MLlib which do?

For comparing results with expected values, I *much* prefer for those values to be in a column in the original input dataset.  That has 2 benefits:
* It makes tests easier to read since inputs + expected values are side-by-side in the code.
* We don't have to worry about Row order.

**Comment by jkbradley** [comment]
Since we are under time pressure, I'm OK with keeping it for now, but let's plan on refactoring the tests later to get rid of the use of globalCheckFunction wherever possible.

**Comment by WeichenXu123** [comment]
@jkbradley 
> When there has been a shuffle, it is likely the Rows will not follow a fixed order.

Agreed. But we can make sure it generate fix order from the last shuffle position in the physical plan RDD lineage. Those model which works like `map` transformation, I think it can make sure output row order to be exactly the same with input row order.

> test statistics (such as min/max ) on global transformer output

This is also used in some tests, such as "predictRaw and predictProbability" testcase in `DecisionTreeClassifierSuite"

> For comparing results with expected values, I much prefer for those values to be in a column in the original input dataset.

Agreed.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala [hunk 1]
```diff
@@ -89,33 +88,31 @@ class DecisionTreeRegressorSuite
...
-    }
 
-    // Since max depth is set to 1, the best split point is that which splits the data
-    // into (0.0, 1.0, 2.0) and (10.0, 12.0, 14.0). The predicted variance for each
-    // data point in the left node is 0.667 and for each data point in the right node
-    // is 2.667
-    val expectedVariances = Array(0.667, 0.667, 0.667, 2.667, 2.667, 2.667)
-    calculatedVariances.zip(expectedVariances).foreach { case (actual, expected) =>
-      assert(actual ~== expected absTol 1e-3)
+    testTransformerByGlobalCheckFunc[(Vector, Double)](varianceDF, dt.fit(varianceDF),
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: The `varianceDF` generated by `TreeTests.setMetadata`, how to add "expected value" column into the DF ? It seems to need some flaky code.  @jkbradley

File: mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala [hunk 2]
```diff
@@ -44,13 +41,11 @@ class IsotonicRegressionSuite
...
-      .transform(dataset)
-      .select("prediction").rdd.map { case Row(pred) =>
-        pred
-      }.collect()
-
-    assert(predictions === Array(1, 2, 2, 2, 6, 16.5, 16.5, 17, 18))
+    testTransformerByGlobalCheckFunc[(Double, Double, Double)](dataset, model,
+      "prediction") { case rows: Seq[Row] =>
+      val predictions = rows.map(_.getDouble(0))
+      assert(predictions === Array(1, 2, 2, 2, 6, 16.5, 16.5, 17, 18))
```
Comment (WeichenXu123) [inline mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala]: Ditto issue.

**Comment by jkbradley** [comment]
> > test statistics (such as min/max ) on global transformer output

> This is also used in some tests, such as "predictRaw and predictProbability" testcase in `DecisionTreeClassifierSuite"

Global statistics actually are not used in that test suite; all of the checks are done row-by-row.

File: mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala (hunk 1 above)
Comment (jkbradley) [reply mllib/src/test/scala/org/apache/spark/ml/regression/DecisionTreeRegressorSuite.scala]: The expected values would have to be added to ```def varianceData```.

File: mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala (hunk 2 above)
Comment (jkbradley) [reply mllib/src/test/scala/org/apache/spark/ml/regression/IsotonicRegressionSuite.scala]: Here, we would have to modify generateIsotonicInput to take the expected values.

**Review by jkbradley — APPROVED** [review]
This LGTM.  I still hold by my comments.  : )  But I'm fine if we do those cleanups later.

Merging with master
Thanks @WeichenXu123 and @MrBago !

**Pull request closed without merging**